addons/portal/wizard/portal_wizard.py

```diff
@@ -70,116 +84,162 @@ class PortalWizardUser(models.TransientModel):
-        if partners_error_empty:
-            error_msg.append("%s\n- %s" % (_("Some contacts don't have a valid email: "),
-                                '\n- '.join(partners_error_empty.mapped('display_name'))))
-        if partners_error_emails:
-            error_msg.append("%s\n- %s" % (_("Several contacts have the same email: "),
-                                '\n- '.join(partners_error_emails.mapped('email'))))
-        if partners_error_user:
-            error_msg.append("%s\n- %s" % (_("Some contacts have the same email as an existing portal user:"),
-                                '\n- '.join([p.email_formatted for p in partners_error_user])))
-        if error_msg:
-            error_msg.append(_("To resolve this error, you can: \n"
-                "- Correct the emails of the relevant contacts\n"
-                "- Grant access only to contacts with unique emails"))
-        return error_msg
-
-    def action_apply(self):
-        self.env['res.partner'].check_access_rights('write')
-        """ From selected partners, add corresponding users to chosen portal group. It either granted
-            existing user, or create new one (and add it to the group).
+
+    user_id = fields.Many2one('res.users', string='User', compute='_compute_user_id')
+    login_date = fields.Datetime(related='user_id.login_date', string='Latest Authentication')
+    is_portal = fields.Boolean('Is Portal', compute='_compute_group_details')
+    is_internal = fields.Boolean('Is Internal', compute='_compute_group_details')
+
+    @api.depends('partner_id')
+    def _compute_user_id(self):
+        for portal_wizard_user in self:
+            user = portal_wizard_user.partner_id.with_context(active_test=False).sudo().user_ids
+            portal_wizard_user.user_id = user[0] if user else False
+
+    @api.depends('user_id', 'user_id.groups_id')
+    def _compute_group_details(self):
+        for portal_wizard_user in self:
+            user = portal_wizard_user.user_id
+
+            if user and user.has_group('base.group_user'):
+                portal_wizard_user.is_internal = True
+                portal_wizard_user.is_portal = False
+            elif user and user.has_group('base.group_portal'):
+                portal_wizard_user.is_internal = False
+                portal_wizard_user.is_portal = True
+            else:
+                portal_wizard_user.is_internal = False
+                portal_wizard_user.is_portal = False
+
+    def action_grant_access(self):
+        """Grant the portal access to the partner.
+
+        If the partner has no linked user, we will create a new one in the same company
+        as the partner (or in the current company if not set).
+
+        An invitation email will be sent to the partner.
+        """
+        self.ensure_one()
+        self._assert_user_email_uniqueness()
+
+        if self.is_portal or self.is_internal:
+            raise UserError(_('This partner already has the portal access.'))
```

Comment:
Maybe add partner name in error so that it is complete ? Even if validation is done per line, always good to have meaningful error messages (if you use it in an automated action for example).

Reply:
Good idea :)